service/history/shard/fx.go

```diff
@@ -48,48 +48,35 @@ var Module = fx.Options(
 	fx.Provide(ShardControllerProvider),
 )
 
-func ShardControllerProvider(
-	config *configs.Config,
-	logger log.Logger,
-	throttledLogger resource.ThrottledLogger,
-	persistenceExecutionManager persistence.ExecutionManager,
-	persistenceShardManager persistence.ShardManager,
-	clientBean client.Bean,
-	historyClient historyservice.HistoryServiceClient,
-	historyServiceResolver membership.ServiceResolver,
-	metricsClient metrics.Client,
-	payloadSerializer serialization.Serializer,
-	timeSource clock.TimeSource,
-	namespaceRegistry namespace.Registry,
-	saProvider searchattribute.Provider,
-	saMapper searchattribute.Mapper,
-	clusterMetadata cluster.Metadata,
-	archivalMetadata archiver.ArchivalMetadata,
-	hostInfoProvider resource.HostInfoProvider,
-) *ControllerImpl {
+type shardControllerDeps struct {
```

Comment:
I'm not clear about further changes after this PR, but I would be concerned regarding adding such a structure. It might reduce code duplication a bit, but it brings in the danger of generating a "god" object that plumbs a lot of dependencies where they do not belong (see the history of resource.Resource).

This change would make more sense if we did not use DI, but with us moving towards uber/fx, it adds another layer of indirection and doesn't fundamentally reduce construction boilerplate.

Reply:
Yes, in some ways it is resource.Resource all over again, but it's a specific one for each package/component. So it's explicitly not a god object. It clearly reduces boilerplate in that I now only have to list a dependency once instead of three times (constructor parameter list, struct field, assignment to struct field). And in this case since shard controller + context are linked and the dependencies are passed on, it was actually six times that's now one. You can argue that the boilerplate is a good thing or a positive trade-off for other things, but it is less code